packages/vite/src/shared/moduleRunnerTransport.ts

```diff
@@ -0,0 +1,312 @@
+import { nanoid } from 'nanoid/non-secure'
+import type { CustomPayload, HotPayload } from 'types/hmrPayload'
+import { promiseWithResolvers } from './utils'
+import type {
+  InvokeMethods,
+  InvokeResponseData,
+  InvokeSendData,
+} from './invokeMethods'
+
+export type ModuleRunnerTransportHandlers = {
+  onMessage: (data: HotPayload) => void
+  onDisconnection: () => void
+}
+
+/**
+ * "send and connect" or "invoke" must be implemented
+ */
+export interface ModuleRunnerTransport {
+  connect?(handlers: ModuleRunnerTransportHandlers): Promise<void> | void
+  disconnect?(): Promise<void> | void
+  send?(data: HotPayload): Promise<void> | void
+  invoke?(
+    data: HotPayload,
+  ): Promise<{ /** result */ r: any } | { /** error */ e: any }>
+  timeout?: number
+}
+
+type InvokeableModuleRunnerTransport = Omit<ModuleRunnerTransport, 'invoke'> &
+  Required<Pick<ModuleRunnerTransport, 'send'>> & {
+    invoke<T extends keyof InvokeMethods>(
+      name: T,
+      data: Parameters<InvokeMethods[T]>,
+    ): Promise<ReturnType<Awaited<InvokeMethods[T]>>>
+  }
+
+const createInvokeableTransport = (
+  transport: ModuleRunnerTransport,
+): InvokeableModuleRunnerTransport => {
+  if (transport.invoke) {
+    const sendOrInvoke = transport.send ?? transport.invoke
+    return {
+      ...transport,
+      async send(data) {
+        await sendOrInvoke(data)
+      },
+      async invoke(name, data) {
+        const result = await transport.invoke!({
+          type: 'custom',
+          event: 'vite:invoke',
+          data: {
+            id: 'send',
+            name,
+            data,
+          } satisfies InvokeSendData,
+        } satisfies CustomPayload)
+        if ('e' in result) {
+          throw result.e
+        }
+        return result.r
+      },
+    }
+  }
+
+  if (!transport.send || !transport.connect) {
+    throw new Error(
+      'transport must implement send and connect when invoke is not implemented',
+    )
+  }
+
+  const rpcPromises = new Map<
+    string,
+    {
+      resolve: (data: any) => void
+      reject: (data: any) => void
+      name: string
+      timeoutId?: ReturnType<typeof setTimeout>
+    }
+  >()
+
+  return {
+    ...transport,
+    connect({ onMessage, onDisconnection }) {
+      transport.connect!({
```

Comment:
Is this promise hanging? I had `send` called before `connect` being resolved in https://github.com/vitejs/vite/pull/18485

```suggestion
      return transport.connect!({
```

Reply:
Should be fixed by 4b349f7827848d8fb2577f99b3ce722700688f84 along with other floating promises.